### ISSUE TITLE
Missing Postgres Support in 7.1+/8.0+

### ISSUE THREAD

**Issue by justjanne** [haiwen/seafile#2586]
Previously, with custom migration scripts it was possible to use Seafile under Postgres.

I’ve written PRs to reintroduce postgres support to the CE, and I’m using this code in production right now with a custom docker image of [seafile](https://git.kuschku.de/justJanne/seafile/-/tree/master/server) and [seahub](https://git.kuschku.de/justJanne/seafile/-/tree/master/seahub).

* https://github.com/haiwen/seafile-server/pull/560
* https://github.com/haiwen/seafdav/pull/66
* https://github.com/haiwen/seahub/pull/5196

It would be great if we could get these PRs merged to get postgres support back into the official release, as I don’t really want to fork seafile in the long-term.

**Comment by KaneTW** [comment]
Bumping this. Would really enjoy postgres support.

**Comment by 98h398hrpohpoai** [comment]
Bumping as well. There's a lot of potential benefit with this inclusion.

**Comment by DenisBessa** [comment]
This would be awesome

**Comment by mabra** [comment]
Would like this also!

**Comment by RWDai** [comment]
Postgres is more power than mysql。I have postgres instance in my server。It's a pity that if Postgres is not supported

**Comment by aigars-smaser** [comment]
+1

**Comment by rcosta-uk** [comment]
+1 
All my applications and self-hosted applications that I run use Postgres.

**Comment by FunkyMuse** [comment]
> +1 All my applications and self-hosted applications that I run use Postgres.

Same, was weird that Seafile dropped support for postgres

**Comment by SR-G** [comment]
Same here, i'd be very interested by a PostGRE compliance for seafile - for the PostGre robustness, and also because Seafile is now the only single app' that i'm running and that still needs MySQL...

**Comment by Zolfax** [comment]
Bumping this.

**Comment by n-rodriguez** [comment]
+1
All my applications and self-hosted applications that I run use Postgres.

**Comment by micheljung** [comment]
[The world is moving away from MySQL/MariaDB, to PostgreSQL](https://trends.google.com/explore?q=%2Fm%2F09gc20r%2C%2Fm%2F05ynw&date=all&geo=Worldwide). It's weird that Seafile went the other direction.

<img width="1027" height="681" alt="Image" src="https://github.com/user-attachments/assets/1deb59cc-179a-4155-a783-082c19895891" />

**Comment by freeplant** [comment]
We know that most newly projects that use PostgreSQL instead MySQL. But Seafile is an old project.

We want to focus on the application layer, instead of spending too much time at the database support layer.

**Comment by n-rodriguez** [comment]
> But Seafile is an old project.

Previously, with custom migration scripts it was possible to use Seafile under Postgres.

**Comment by freeplant** [comment]
There are now too many components in Seafile, including the newly added SeaDoc, metadata server, and seafile-ai.

Currently, for each major release, we require three people and two months to prepare: one person handles packaging and upgrading the corresponding scripts for configuration changes, another focuses on updating the documentation, and the third is responsible for testing. Despite this effort, we still encounter numerous corner cases related to Seafile installation and configuration after the release.

Ensuring that all components support both MySQL and Postgres will nearly double our maintenance workload.

**Comment by justjanne** [comment]
It wouldn't be nearly as much effort if you had an actual, cleanly abstracted storage backend, instead of just #ifdef's and puzzled-together SQL queries.

With a good architecture, I'd be able to just maintain & update a patchset that adds postgres support back, but with the codebase (at least in 7, 8, 9 and 10, haven't checked 11 yet) that's almost impossible.

In the past you not only had MySQL and Postgres, but also Oracle for enterprise customers from what I remember?

**Comment by freeplant** [comment]
> In the past you not only had MySQL and Postgres, but also Oracle for enterprise customers from what I remember?

In the past, we support MySQL, Postgres, SQLite3 and Oracle. Oracle support was removed a long time ago. SQLite3 support was removed in version 11.0

**Comment by justjanne** [comment]
@freeplant I'll be very direct here: Your release process is slow because of tech debt. Seafile is what we'd call "spaghetti code" in english. But the databases you support have nothing to do with that, and if you clean up the tech debt, adding features will be much faster, and support multiple databases will be much easier.

I've now started adding postgres support back into 13.0, while at the same time [cleaning up the spaghetti code](https://github.com/justjanne/seafile-server/commit/8d79bfe6eb4d2cf2b09609d160962fbf6d5956c6), making it much easier to [build and run the entire seafile stack locally from source](https://github.com/justjanne/seafile-stack), and creating a more robust test setup.

**Comment by bmurphy1976** [comment]
I'm curious to know why a major release takes 3 months.  I'm sure sloppy code contributes BUT that also sounds like a big organizational/process bottleneck.  Some of it is probably for good reasons, but coupling that with the db layer limitations does sound like concerns are getting mixed up.

**Comment by justjanne** [comment]
> I'm curious to know why a major release takes 3 months.  I'm sure sloppy code contributes BUT that also sounds like a big organizational/process bottleneck.

Seafile is split into many different projects, spread across multiple repositories, written in different languages, all of which mix database interactions and business logic together.

Looking at the code and docs, it seems to be a fully manual release process, with very little automation or testing.

That's a perfect recipe for complicated and hard to fix bugs. It's no wonder the release process is this slow and unreliable.

**Comment by justjanne** [comment]
@bmurphy1976 Here's an example of business logic and SQL being somewhat interwoven
https://github.com/haiwen/seafile-server/blob/377d6e5805add1eaffc25eeb83a1b5310eeb3131/common/group-mgr.c#L758-L851

Before anyone asks about the new go fileserver, that's actually the exact same logic:
https://github.com/haiwen/seafile-server/blob/377d6e5805add1eaffc25eeb83a1b5310eeb3131/fileserver/share/share.go#L101-L151

(note that it's generally bad practice to dynamically create SQL statements in this way, there are much better and cleaner alternatives available)

**Comment by bmurphy1976** [comment]
@justjanne I agree that's not the best code layout but it doesn't explain the slow release process.  I think you are right though with your intuition about automation and testing.  That's far more likely the culprit in my opinion.

**Comment by n-rodriguez** [comment]
> (note that it's generally bad practice to dynamically create SQL statements in this way, there are much better and cleaner alternatives available)

especially in Go 😢

**Comment by micheljung** [comment]
While I agree with the points being made, I'd like to raise awareness that this is an open source project ran by volunteers, and we can all be very grateful for their work. "Your code sucks" won't help them stay motivated. Helping out, however, might 🩵

**Comment by justjanne** [comment]
> I'd like to raise awareness that this is an open source project ran by volunteers

Not really. We should still be grateful, but Seafile is a commercial product by Seafile Ltd., a chinese corporation (formerly chinese/german joint venture), which sells a paid "pro" version with added functionality.

I believe the pro versions are also why they've never merged a community PR in this repository, as that'd bring the community edition and pro version out of sync.

That's why I believe that if we want any changes to be merged, we'll have to create a community-driven alternative instead.

**Comment by iWantToKeepAnon** [comment]
Please bring either psql or sqlite back plz
